text/0000-dyn-trait-syntax.md

```diff
@@ -0,0 +1,102 @@
+
+enums and impl Trait simply don't have these costs. It's strange that the more concise syntax gives you a feature that is often slower and rarely faster than its alternatives.
+
+### favors a feature that ... often cannot be used at all when its alternatives can
+
+Many traits simply can't have trait objects at all, because they don't meet the object safety rules.
+
+In contrast, impl Trait and generics work with any trait. It's strange that the more concise syntax gives you the feature that's least likely to compile.
+
+### impl Trait is going to require a significant shift in idioms and teaching materials all on its own
+
+Today, when you want to return a type implementing a trait without writing out the type's name, you typically `Box` a trait object and accept the potential runtime cost. This includes most functions that return closures, iterators, futures, or combinations thereof. Most of those functions should switch to impl Trait once that syntax stabilizes and becomes the preferred idiomatic way of doing this, including many public API methods.
+
+The way we teach the trait system will also have to change to describe impl Trait alongside all the existing ways of using traits via generics and trait objects, and explain when impl Trait is preferable to those and other options like enums. Moreover, the way we teach closures, iterators and futures will likely need to mention why impl Trait is useful for those types and use impl Trait in many examples, as well as when impl Trait isn't enough and you do need dyn Trait after all.
+
+Ideally, introducing dyn Trait won't create much additional churn on top of impl Trait, since these idiom shifts and documentation rewrites can account for both of those changes together.
+
+### "dyn Trait vs impl Trait" is much nicer for teaching and ergonomics than "bare trait vs impl Trait"
+
+There's a natural parallel between the impl/dyn keywords and static/dynamic dispatch that we'll likely mention in The Book. Having a keyword for both kinds of dispatch correctly implies that both are important and choosing between the two is often non-trivial, while today's syntax may give the incorrect impression that trait objects are the default and impl Trait is a more niche feature.
+
+After impl Trait stabilizes, it will become more common to accidentally write a trait object without realizing it by forgetting the impl keyword. This often leads to unhelpful and cryptic errors about your trait not implementing Sized. With a switch to dyn Trait, these errors could become as simple and self-evident as "expected a type, found a trait, did you mean to write impl Trait?".
+
+# Explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The functionality of `dyn Trait` is identical to today's trait object syntax.
```

Comment:
It might be good to mention the `impl MyTrait for dyn OtherTrait {}` syntax here, since it directly addresses the `impl MyTrait for OtherTrait {}` issue in the "Motivation" section.